### PR TITLE
[Add] `TimmWrapper`

### PR DESCRIPTION
**Pull request by ariG23498**
This PR adds a blog post on the latest `timm` integration with the `transformers` ecosystem.

CC: @rwightman @qubvel @pcuenca @Vaibhavs10

### CODE REVIEW

File: timmwrapper.md [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+> Check out the full repository for all code examples and notebooks:
+> 🔗 [TimmWrapper Examples](https://github.com/ariG23498/timmwrapper-examples)  
+
+## What is timm?
+
+The [PyTorch Image Models (`timm`) library](https://huggingface.co/docs/timm/en/index)
+offers a rich collection of state-of-the-art computer vision models,
+along with useful layers, utilities, optimizers, and data augmentations.
+With more than 32K GitHub stars, more than 200K daily downloads at the time of writing,
+it’s a go-to resource for image classification, object detection, segmentation, and more.
```
Comment (qubvel) [inline timmwrapper.md:42]: While `timm` offers state-of-the-art image classification models that can be used as building blocks for object detection and segmentation, I'm not sure if this sounds correct since, as far as I know, `timm` itself does not have object detection/segmentation models yet.

File: timmwrapper.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+- 🎯 **Fine-Tuning with Trainer API**: Fine-tune `timm` models using the `Trainer` API and even integrate with adapters like low rank adaptation (LoRA).
+- 🚀 **Torch Compile for Speed**: Leverage `torch.compile` to optimize inference time.
+
+## Pipeline API: Using timm Models for Image Classification
+
+One of the standout features of the `timm` integration is that it allows you to leverage the 🤗 **`pipeline` API**.
+The **`pipeline` API** abstracts away a lot of complexity, making it easy to load a pre-trained model,
+perform inference, and view results with a few lines of code.
+
+Let’s see how to use a transformers pipeline with the *MobileNetV4* (not available in `transformers`) model from `timm`:
```
Comment (qubvel) [inline timmwrapper.md:65]: The phrase "not available in `transformers`" sounds contradictory to me. It might be better to say something like "became available in `transformers` with `timm`", WDYT?

File: timmwrapper.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+Label: tiger cat            Score: 0.21
+Label: Egyptian cat         Score: 0.02
+Label: bee                  Score: 0.00
+Label: marmoset             Score: 0.00
+```
+
+## Gradio Integration: Building a Food Classifier Demo 🍣  
+
+Want to quickly create an interactive web app for image classification? **Gradio** makes it simple
+to build a user-friendly interface with minimal code. Let's combine **Gradio** with the `pipeline` API
```
Comment (qubvel) [inline timmwrapper.md:99]: ❤️

File: timmwrapper.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+# Benchmark function
+def run_benchmark(model, input_data, runs=300):
+    model.eval()
+    with torch.no_grad():
+        times = [time.time() - time.time() for _ in range(runs)]
+        for i in range(runs):
+            start = time.time()
+            _ = model(**input_data)
+            times[i] = time.time() - start
+    return sum(times) / runs
```
Comment (qubvel) [inline timmwrapper.md:349]: This way, it looks more pythonic

```suggestion
        times = []
        for i in range(runs):
            start = time.time()
            _ = model(**input_data)
            times.append(time.time() - start)
    return sum(times) / runs
```

Comment (qubvel) [reply timmwrapper.md:349]: Do we need a warm-up, or is it intended to be omitted?

**Review by qubvel — COMMENTED** [review]
Super good! Great job, looking forward to it being published!

File: timmwrapper.md (hunk 1 above)
Comment (ariG23498) [reply timmwrapper.md:42]: Do you think I should change the sentence to better communicate this point? `timm` is widely used for a lot of tasks other than image classification, due to the backbones. I am not sure how to place this sentiment better though.

Do you have something in mind?

File: timmwrapper.md (hunk 2 above)
Comment (ariG23498) [reply timmwrapper.md:65]: I do want to emphasise that MobileNetV4 is not available in the `transformers` repository. The wrapper makes it available, and there are many such models.

Can we adjust the statement to "does not have a transformers implementation, but now has became available in transformers with timm". I think this statement drives the point home, what do you think?

File: timmwrapper.md (hunk 4 above)
Comment (ariG23498) [reply timmwrapper.md:349]: This is intended to be omitted. The end to end notebooks are all [in this repository](https://github.com/ariG23498/timm-wrapper-examples).

File: timmwrapper.md (hunk 1 above)
Comment (qubvel) [reply timmwrapper.md:42]: Agree, maybe it's just my perception, let's see.

File: timmwrapper.md (hunk 2 above)
Comment (qubvel) [reply timmwrapper.md:65]: Sounds good

File: timmwrapper.md (hunk 1 above)
Comment (rwightman) [reply timmwrapper.md:42]: @ariG23498 What about `it's a go-to resource for image classification and feature extraction for object detection, segmentation, image search, and other downstream tasks.`

File: timmwrapper.md [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+            start = time.time()
+            _ = model(**input_data)
+            times.append(time.time() - start)
+    return sum(times) / runs
+
+# Run benchmarks
+time_no_compile = run_benchmark(model, processed_input)
+compiled_model = torch.compile(model).to(device)
+time_compile = run_benchmark(compiled_model, processed_input)
+
```
Comment (rwightman) [inline timmwrapper.md:356]: FWIW This benchmark isn't technically correct if it's using an accelerator. 

A synchronize call is needed after the model forward ... e.g. for cuda `torch.cuda.synchronize(device=device)`

Otherwise it's just timing the asynchronous dispatch. Also `time.perf_counter()` is a bit better to use here.

Comment (ariG23498) [reply timmwrapper.md:356]: That is a great catch!

I have made the following changes, does it make sense now?
```py
def run_benchmark(model, input_data, warmup_runs=5, benchmark_runs=300):
    # Warm-up phase
    model.eval()
    with torch.no_grad():
        for _ in range(warmup_runs):
            _ = model(**input_data)

    # Benchmark phase
    times = []
    with torch.no_grad():
        for _ in range(benchmark_runs):
            start_time = time.perf_counter()
            _ = model(**input_data)
            if device.type == "cuda":
                torch.cuda.synchronize(device=device)  # Ensure synchronization for CUDA
            times.append(time.perf_counter() - start_time)

    avg_time = sum(times) / benchmark_runs
    return avg_time
```

Comment (rwightman) [reply timmwrapper.md:362]: @ariG23498 yup, looks good!

File: timmwrapper.md
Comment (pcuenca) [inline timmwrapper.md:1]: how about rename to timm-transformers ?

File: timmwrapper.md [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+full 'round-trip' compatibility. Namely, using the wrapper one can fine-tune a timm model on a new
+dataset using `transformer`'s `Trainer`, publish the resulting model to the Hugging Face hub,
+and then load the fine-tuned model in `timm` again using `timm.create_model('hf-hub:my_org/my_fine_tuned_model', pretrained=True)`.
+
+## What is timm?
+
+The [PyTorch Image Models (`timm`) library](https://huggingface.co/docs/timm/en/index)
+offers a rich collection of state-of-the-art computer vision models,
+along with useful layers, utilities, optimizers, and data augmentations.
+With more than 32K GitHub stars, more than 200K daily downloads at the time of writing,
```
Comment (pcuenca) [inline timmwrapper.md:46]:
```suggestion
With more than 32K GitHub stars and more than 200K daily downloads at the time of writing,
```

File: timmwrapper.md [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+it's a go-to resource for image classification and feature extraction for object detection,
+segmentation, image search, and other downstream tasks.
+
+With pre-trained models covering a wide range of architectures, `timm` simplifies the workflow for
+computer vision practitioners.
+
+## Why Use the TimmWrapper?
```
Comment (pcuenca) [inline timmwrapper.md:53]:
```suggestion
## Why Use the timm integration?
```

File: timmwrapper.md [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+
+While 🤗 `transformers` supports several vision models, `timm` offers an even broader collection,
+including many mobile-friendly and efficient models not available in transformers.
+
+The `timm` integration bridges this gap, bringing the best of both worlds:
+- ✅ **Pipeline API Support**: Easily plug any `timm` model into the high-level `transformers` pipeline for streamlined inference.
+- 🧩 **Compatibility with Auto Classes**: While `timm` models aren’t natively compatible with `transformers`, the integration makes them work seamlessly with the `Auto` classes API.
+- ⚡ **Quick Quantization**: With just ~5 lines of code, you can quantize **any** `timm` model for efficient inference.
+- 🎯 **Fine-Tuning with Trainer API**: Fine-tune `timm` models using the `Trainer` API and even integrate with adapters like low rank adaptation (LoRA).
+- 🚀 **Torch Compile for Speed**: Leverage `torch.compile` to optimize inference time.
```
Comment (pcuenca) [inline timmwrapper.md:63]:
```suggestion
- 🔁 **Round trip to timm**: Use fine-tuned models back in `timm`.
- 🚀 **Torch Compile for Speed**: Leverage `torch.compile` to optimize inference time.
```

File: timmwrapper.md [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+
+## Pipeline API: Using timm Models for Image Classification
+
+One of the standout features of the `timm` integration is that it allows you to leverage the 🤗 **`pipeline` API**.
+The **`pipeline` API** abstracts away a lot of complexity, making it easy to load a pre-trained model,
+perform inference, and view results with a few lines of code.
+
+Let's see how to use a transformers pipeline with the *MobileNetV4* (does not have a `transformers` implementation,
+but now has became available in `transformers` with `timm`) model from `timm`:
```
Comment (pcuenca) [inline timmwrapper.md:72]:
```suggestion
Let's see how to use a transformers pipeline with the *MobileNetV4*. This architecture does not have a native `transformers` implementation, but can be easily used from `timm`:
```

File: timmwrapper.md [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+``` 
+
+Quantized models perform **almost identically** to full-precision models during inference:
+
+| **Model**         | **Label**                | **Accuracy** |
+|-------------------|-------------------------|-----------|
+| Original Model    | remote control, remote   | 0.35%     |
+| Quantized Model   | remote control, remote   | 0.33%     |
+
+## Supervised Fine-Tuning with `TimmWrapper`
```
Comment (pcuenca) [inline timmwrapper.md:217]:
```suggestion
## Supervised Fine-Tuning of `timm` models
```

File: timmwrapper.md [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+    ignore_mismatched_sizes=True,
+)
+inference_model = PeftModel.from_pretrained(model, repo_name)
+
+# Make prediction with the model
+```
+
+![image of sushi with prediction from a fine tuned model](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/timmwrapper/prediction-food.png)
+
+## Torch Compile: Instant Speedup 🚀
```
Comment (pcuenca) [inline timmwrapper.md:336]: I'd add a short section about inference with timm if we can

File: timmwrapper.md [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+# Results
+print(f"Without torch.compile: {time_no_compile:.4f} s")
+print(f"With torch.compile: {time_compile:.4f} s")
+```
+
+![compile timing](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/timmwrapper/compile.png)
+
+## Wrapping Up
+
+`timm` integrated into the Hugging Face ecosystem opens new doors for leveraging state-of-the-art vision models
```
Comment (pcuenca) [inline timmwrapper.md:382]:
```suggestion
`timm`'s integration with transformers opens new doors for leveraging state-of-the-art vision models
```

File: timmwrapper.md [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+with minimal effort. Whether you're looking to fine-tune, quantize, or simply run inference, this
+integration provides a unified API to streamline your workflow.
+
+By combining timm's model repository with the extensive ecosystem of transformers, you get the best
+of both worlds. Start exploring today and unlock new possibilities in computer vision!
```
Comment (pcuenca) [inline timmwrapper.md:387]:
```suggestion
Start exploring today and unlock new possibilities in computer vision!
```

A bit repetitive with the above, maybe?

File: timmwrapper.md [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+> ```
+
+> [!NOTE]  
+> Check out the full repository for all code examples and notebooks:
+> 🔗 [TimmWrapper Examples](https://github.com/ariG23498/timmwrapper-examples)  
+
+Also, worth mentioning, and one of Ross' (creator of `timm`) favourite features, this integration maintains
+full 'round-trip' compatibility. Namely, using the wrapper one can fine-tune a timm model on a new
+dataset using `transformer`'s `Trainer`, publish the resulting model to the Hugging Face hub,
+and then load the fine-tuned model in `timm` again using `timm.create_model('hf-hub:my_org/my_fine_tuned_model', pretrained=True)`.
```
Comment (pcuenca) [inline timmwrapper.md:39]: Not sure this is the best place for this, may not be too visible after the notes. What do you think about adding it as an additional bullet point to the list above?

**Review by pcuenca — APPROVED** [review]
Looks great! Suggested a few nits, removed a couple of additional `TimmWrapper` mentions (there are still some, so it should be fine) and suggested to rename the file so the url path is `/blog/timm-transformers` rather than `/blog/timmwrapper`.

File: timmwrapper.md [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+
+**Model Example:**  
+LoRA Fine-Tuned ViT on **Food-101**: [**`vit_base_patch16_224.augreg2_in21k_ft_in1k.lora_ft_food101`**](https://huggingface.co/ariG23498/vit_base_patch16_224.augreg2_in21k_ft_in1k.lora_ft_food101)
+
+
+LoRA is just one example of efficient adapter-based fine-tuning methods you can apply to `timm` models.
+The integration of `timm` with the 🤗 ecosystem opens up a wide variety of **parameter-efficient fine-tuning (PEFT)** techniques,
+allowing you to choose the method that best fits your use case.
+
+### Inference with LoRA Fine-Tuned Model
```
Comment (pcuenca) [inline timmwrapper.md:311]: I think the lora fine-tunes cannot be used back in `timm`, can they?

File: timmwrapper.md
Comment (ariG23498) [reply timmwrapper.md:1]: Done!

File: timmwrapper.md (hunk 11 above)
Comment (ariG23498) [reply timmwrapper.md]: Done!

File: timmwrapper.md (hunk 14 above)
Comment (ariG23498) [reply timmwrapper.md]: I have moved this to a later section (which is dedicated to this point)

File: timm-transformers.md [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+
+# Timm ❤️ Transformers: Use any timm model with transformers
+
+Ever wanted to leverage the vast collection of state-of-the-art models from the
+[`timm` library](https://huggingface.co/models?library=timm&sort=trending) using
+the familiar 🤗 `transformers` ecosystem?
+
+Enter the [`TimmWrapper`](https://huggingface.co/docs/transformers/main/en/model_doc/timm_wrapper)—a simple,
+yet powerful tool that unlocks this potential.
+
```
Comment (Vaibhavs10) [inline timm-transformers.md:20]: I'd maybe contextualise this a bit more - the start feels quite a bit dry IMO - In a line we can put all the things the integration brings to timm models - compile, quantisation etc, just to give people a sneak peek.

It's not clear from the current first couple lines.

**Review by Vaibhavs10 — APPROVED** [review]
Lovely read, just one comment on the start of the blogpost

File: timm-transformers.md (hunk 16 above)
Comment (ariG23498) [reply timm-transformers.md:19]: I completely agree with your comment.
Do you think the current version is better?

Comment (Vaibhavs10) [reply timm-transformers.md:19]: Sounds good to me!
